modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/StatisticsGatherer.java

```diff
@@ -272,11 +272,18 @@ public void stop() {
         if (log.isTraceEnabled())
             log.trace(String.format("Statistics gathering stopping %d task...", gatheringInProgress.size()));
 
-        gatheringInProgress.values().forEach(ctx -> ctx.futureGather().cancel(true));
-
-        gatheringInProgress.clear();
+        cancelAllTasks();
 
         if (log.isDebugEnabled())
             log.debug("Statistics gathering stopped.");
     }
+
+    /**
+     * Cancel all currently running statistics gathering tasks.
+     */
+    public void cancelAllTasks() {
+        gatheringInProgress.values().forEach(ctx -> ctx.futureGather().cancel(true));
+
+        gatheringInProgress.clear();
+    }
```

Comment:
This can be is called from StatisticsManager.disableOperations() as a reaction on configuration changes.
And StatisticsManager.processObsolescence() that is called from scheduler can put  into gatheringInProgress in parallel.
Seems, both StatisticsManager methods are marked synchronized, but actually don't check the state under lock.

Reply:
It won't totally fix the possibility of failure cause we can change state during processObsolescence execution. But I move state check directly into it to improve readability and lower this chance a bit.